File: src/types/campaign.ts
export interface CampaignSnapshot {
  resources: number;
  wave_number: number;
  owned_region_ids: number[];
  regions: RegionSnapshot[];
  available_actions: AvailableAction[];
  tech_tree: TechTreeSnapshot;
  wave_income?: number;
}

export interface TechTreeSnapshot {
  unlocked_types: string[];
  upgrades: TypeUpgradeSnapshot[];
}

export interface TypeUpgradeSnapshot {
  interceptor_type: string;
  thrust_level: number;
  yield_level: number;
  guidance_level: number;
}

export interface RegionSnapshot {
  id: number;
  name: string;
  terrain: string;
  owned: boolean;
  expandable: boolean;
  cities: CitySnapshotCampaign[];
  battery_slots: BatterySlotSnapshot[];
  map_x: number;
  map_y: number;
  expansion_cost: number;
}

export interface CitySnapshotCampaign {
  x: number;
  y: number;
  population: number;
  health: number;
  max_health: number;
}

export interface BatterySlotSnapshot {
  x: number;
  y: number;
  occupied: boolean;
  ammo: number | null;
  max_ammo: number | null;
}

export type AvailableAction =
  | { ExpandRegion: { region_id: number; cost: number } }
  | { PlaceBattery: { region_id: number; slot_index: number; cost: number } }
  | { RestockBattery: { region_id: number; slot_index: number; cost: number } }
  | { RepairCity: { region_id: number; city_index: number; cost: number; health_to_restore: number } }
  | { UnlockInterceptor: { interceptor_type: string; cost: number; min_wave: number } }
  | { UpgradeInterceptor: { interceptor_type: string; axis: string; cost: number; current_level: number } }
  | "StartWave";

File: src-tauri/src/commands/campaign.rs
use crate::engine::game_loop::{EngineCommand, GameEngine};

#[tauri::command]
pub fn start_wave(engine: tauri::State<'_, GameEngine>) {
    engine.send_command(EngineCommand::StartWave);
}

#[tauri::command]
pub fn continue_to_strategic(engine: tauri::State<'_, GameEngine>) {
    engine.send_command(EngineCommand::ContinueToStrategic);
}

#[tauri::command]
pub fn expand_region(engine: tauri::State<'_, GameEngine>, region_id: u32) {
    engine.send_command(EngineCommand::ExpandRegion { region_id });
}

#[tauri::command]
pub fn place_battery(engine: tauri::State<'_, GameEngine>, region_id: u32, slot_index: u32) {
    engine.send_command(EngineCommand::PlaceBattery {
        region_id,
        slot_index,
    });
}

#[tauri::command]
pub fn restock_all_batteries(engine: tauri::State<'_, GameEngine>) {
    engine.send_command(EngineCommand::RestockAllBatteries);
}

#[tauri::command]
pub fn repair_city(engine: tauri::State<'_, GameEngine>, city_index: u32) {
    engine.send_command(EngineCommand::RepairCity { city_index });
}

#[tauri::command]
pub fn unlock_interceptor(engine: tauri::State<'_, GameEngine>, interceptor_type: String) {
    engine.send_command(EngineCommand::UnlockInterceptor { interceptor_type });
}

#[tauri::command]
pub fn upgrade_interceptor(engine: tauri::State<'_, GameEngine>, interceptor_type: String, axis: String) {
    engine.send_command(EngineCommand::UpgradeInterceptor { interceptor_type, axis });
}

#[tauri::command]
pub fn get_campaign_state(engine: tauri::State<'_, GameEngine>) {
    engine.send_command(EngineCommand::GetCampaignState);
}

#[tauri::command]
pub fn new_game(engine: tauri::State<'_, GameEngine>) {
    engine.send_command(EngineCommand::NewGame);
}
